src/haz3lweb/view/ScratchMode.re

```diff
@@ -98,7 +98,220 @@ let download_slide_state = state => {
+  let editor = Editors.get_editor(editors);
+  let zipper = editor.state.zipper;
+  let unselected = Zipper.unselect_and_zip(zipper);
+  let (term, _) = MakeTerm.go(unselected);
+  let info_map = Statics.mk_map(term);
+  switch (zipper.backpack, Indicated.index(zipper)) {
+  | ([_, ..._], _) => [div([text("")])]
+  | (_, None) => [div([text("")])]
+  | (_, Some(id)) =>
+    switch (Id.Map.find_opt(id, info_map)) {
+    | None => [div([text("")])]
+    | Some(ci) =>
+      let view_fun = (name, ids, _) => [
+        div(
+          ~attr=
+            Attr.many([
+              clss(["breadcrumb_bar_function"]),
+              Attr.on_click(_ =>
+                inject(
+                  UpdateAction.PerformAction(Jump(TileId(List.hd(ids)))),
+                )
+              ),
+            ]),
+          [text(name)],
+        ),
+      ];
+      let rec tag_term = (term: TermBase.UExp.t, level: int) => {
+        switch (term.term) {
+        | Let({term: Var(name), _}, {term: Fun(_, fbody), _}, body) =>
+          let l1 = tag_term(body, level);
+          let l2 = tag_term(fbody, level + 1);
+          List.cons((level, (name, term.ids)), l1) @ l2;
+        | Module(
+            {term: TypeAnn({term: Constructor(name), _}, _), _},
+            t1,
+            t2,
+          )
+        | Module({term: Constructor(name), _}, t1, t2) =>
+          let l1 = tag_term(t2, level);
+          let l2 = tag_term(t1, level + 1);
+          List.cons((level, (name, term.ids)), l1) @ l2;
+        | UnOp(_, t1)
+        | Fun(_, t1)
+        | TyAlias(_, _, t1)
+        | Test(t1)
+        | Parens(t1)
+        | Ap(_, t1) => tag_term(t1, level)
+        | Let(_, t1, t2)
+        | Seq(t1, t2)
+        | Cons(t1, t2)
+        | ListConcat(t1, t2)
+        | BinOp(_, t1, t2) =>
+          let l1 = tag_term(t1, level);
+          let l2 = tag_term(t2, level);
+          l1 @ l2;
+        | If(t1, t2, t3) =>
+          tag_term(t1, level) @ tag_term(t2, level) @ tag_term(t3, level)
+        | ListLit(t)
+        | Tuple(t) => List.concat(List.map(t1 => tag_term(t1, level), t))
+        | Match(t1, t) =>
+          let l1 = tag_term(t1, level);
+          let l2 =
+            List.concat(List.map(t1 => tag_term(snd(t1), level), t));
+          l1 @ l2;
+        | _ => []
+        };
+      };
+      let combineList = (lst: list((int, (string, list(Uuidm.t))))) => {
+        let combine = (acc, (key, id)) =>
+          switch (List.assoc_opt(key, acc)) {
+          | Some(ids) =>
+            List.remove_assoc(key, acc) @ [(key, [id, ...ids])]
+          | None => [(key, [id]), ...acc]
+          };
+        let result = List.fold_left(combine, [], lst);
+        List.map(((key, ids)) => (key, List.rev(ids)), result);
+      };
+      let tagged = tag_term(term, 1);
+      let lst = combineList(tagged);
+      let ancestors = Info.ancestors_of(ci);
+      let rec filter_ancestors = (ancestors_lst: Info.ancestors, level: int) =>
```

Comment:
This function is very ugly, but I'm not sure whether there is a better way...

Reply:
sure, here's a couple tricks that might help:

1. you can combine patterns when pattern matching
2. you can add a "when" to pattern matches as well

so this whole thing could probably become just one pattern match with four cases

```reason
switch (Id.Map.find_opt(List.hd(ancestors_lst), info_map)) {
  | Some(Info.InfoExp({Fun(_), ancestors, _})) when List.length(ancestors) >= 1 => [ ... ]
  | ...
```